tlatools/org.lamport.tlatools/src/tlc2/tool/fp/LongArray.java

```diff
@@ -1,310 +1,209 @@
+ *    indexing; this limits size to approximately 2 billion elements, which
+ *    TLC can easily exceed. By allocating memory ourselves we can use a long
+ *    for both capacity and indexing.
+ * 2. We want to allocate this memory permanently, to ensure it is never
+ *    garbage-collected.
+ *
+ * Before using the newer java.lang.foreign and java.lang.invoke Foreign
+ * Function & Memory (FFM) APIs, this class used sun.misc.Unsafe. That API
+ * has been deprecated; see https://openjdk.org/jeps/471
+ *
+ * Code & understanding adapted from the following StackOverflow questions:
+ *  - Allocating MemorySegment: https://stackoverflow.com/q/78587689/2852699
+ *  - How to CAS with VarHandle: https://stackoverflow.com/q/78599864/2852699
+ *  - On layout alignment: https://stackoverflow.com/q/78624634/2852699
+ *  - 32-bit compatibility: https://stackoverflow.com/q/78625226/2852699
  */
 public final class LongArray {
 
-	private final sun.misc.Unsafe unsafe;
-	
 	/**
-	 * The base address of this direct memory allocated with Unsafe.
+	 * Segment holding the allocated memory.
+	 */
+	private final MemorySegment segment;
+
+	/**
+	 * The layout of the allocated memory.
 	 */
-	private final long baseAddress;
+	private final ValueLayout.OfLong layout = ValueLayout.JAVA_LONG;
+
+	/**
+	 * The handle to use for atomic compare-and-set (CAS) operations.
+	 */
+	private final VarHandle handle;
 
 	/**
 	 * Maximum number of elements that can be contained in this array.
 	 */
-	private final long length;
+	private final long capacity;
 	
 	/**
-	 * CHOOSE logAddressSize \in 1..(Long.SIZE / 8): 2^logAddressSize = (Long.SIZE / 8)
+	 * Constructs a zeroed array of the given capacity.
+	 *
+	 * @param capacity The maximum number of longs the array can hold.
 	 */
-	private static final int logAddressSize = 3;
+	LongArray(final long capacity) {
+		this.capacity = capacity;
 
-	LongArray(final long positions) {
-		this.length = positions;
-		this.unsafe = getUnsafe();
-		
-		// LongArray is only implemented for 64bit architectures. A 32bit
-		// implementation might be possible. However, implementing CAS (see
-		// trySet) seems impossible when values have to be split in two
-		// parts (low/hi) on a 32 bit architecture.
-		// addressSize(): Report the size in bytes of a native pointer, as
-		// stored via #putAddress . This value will be either 4 or 8. We
-		// expect 8 (Long.SIZE / 8) which is the size of a TLC fingerprint
-		// (see FP64).
-		Assert.check(this.unsafe.addressSize() == (Long.SIZE / 8), EC.GENERAL);
-		baseAddress = this.unsafe.allocateMemory(positions << logAddressSize);
+		assert isSupported() : "LongArray requires a 64-bit system architecture.";
+
+		// This allocates the memory in the global arena, which ensures it will
+		// never be garbage-collected and thus guarantees a memory leak. Allocated
+		// memory is zeroed automatically.
+		this.segment = Arena.global().allocate(this.layout, capacity);
```

Comment:
https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/Arena.html#thread-confinement suggests a potential performance penalty.  It should be confirmed that `global` provides comparable performance to Unsafe.